executor/executor.go

```diff
@@ -461,21 +461,104 @@ func (e *TableDualExec) Close() error {
 	return nil
 }
 
+func substituteCorCol2Constant(expr expression.Expression) (expression.Expression, bool) {
+	switch x := expr.(type) {
+	case *expression.ScalarFunction:
+		allConstant := true
+		newArgs := make([]expression.Expression, 0, len(x.GetArgs()))
+		for _, arg := range x.GetArgs() {
+			newArg, ok := substituteCorCol2Constant(arg)
+			newArgs = append(newArgs, newArg)
+			allConstant = allConstant && ok
+		}
+		if allConstant {
+			val, _ := x.Eval(nil)
+			return &expression.Constant{Value: val}, true
+		}
+		var newSf expression.Expression
+		if x.FuncName.L == ast.Cast {
+			newSf = expression.NewCastFunc(x.RetType, newArgs[0], x.GetCtx())
+		} else {
+			newSf, _ = expression.NewFunction(x.GetCtx(), x.FuncName.L, x.GetType(), newArgs...)
+		}
+		return newSf, false
+	case *expression.CorrelatedColumn:
+		return &expression.Constant{Value: *x.Data, RetType: x.GetType()}, true
+	case *expression.Constant:
+		return x.Clone(), true
+	default:
+		return x.Clone(), false
+	}
+}
+
 // SelectionExec represents a filter executor.
 type SelectionExec struct {
 	Src       Executor
 	Condition expression.Expression
 	ctx       context.Context
 	schema    *expression.Schema
+
+	// scanController will tell whether this selection need to
+	// control the condition of below scan executor.
+	scanController      bool
+	controllerInit      bool
+	accessConditions    []expression.Expression
+	idxFilterConditions []expression.Expression
+	tblFilterConditions []expression.Expression
 }
 
 // Schema implements the Executor Schema interface.
 func (e *SelectionExec) Schema() *expression.Schema {
 	return e.schema
 }
 
+// initController will init the conditions of the below scan executor.
+// It will first substitute the correlated column to constant, then calc the range by new condition.
+func (e *SelectionExec) initController() error {
+	sc := e.ctx.GetSessionVars().StmtCtx
+	client := e.ctx.GetClient()
+	accesses := make([]expression.Expression, 0, len(e.accessConditions))
+	for _, cond := range e.accessConditions {
+		newCond, _ := substituteCorCol2Constant(cond)
+		accesses = append(accesses, newCond)
+	}
+	switch x := e.Src.(type) {
+	case *XSelectTableExec:
+		ranges, err := plan.BuildTableRange(accesses, sc)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		x.ranges = ranges
+		tblFilters := make([]expression.Expression, 0, len(e.tblFilterConditions))
+		for _, cond := range e.tblFilterConditions {
+			newCond, _ := substituteCorCol2Constant(cond)
+			tblFilters = append(tblFilters, newCond)
+		}
+		x.where, _, _ = plan.ExpressionsToPB(sc, tblFilters, client)
+		return nil
+	case *XSelectIndexExec:
+		x.indexPlan.AccessCondition = accesses
+		err := plan.BuildIndexRange(sc, x.indexPlan)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		x.indexPlan.IndexConditionPBExpr, _, _ = plan.ExpressionsToPB(sc, e.idxFilterConditions, client)
```

Comment:
`e.idxFilterCondtion` use the value of `*Select.IdxConditions` directly, i think this don't need to update in executor phase.